libwasmvm/src/error/go.rs

```diff
@@ -15,77 +15,185 @@ use crate::memory::UnmanagedVector;
 //
 #[repr(i32)] // This makes it so the enum looks like a simple i32 to Go
 #[derive(PartialEq)]
-pub enum GoResult {
-    Ok = 0,
+pub enum GoError {
+    None = 0,
     /// Go panicked for an unexpected reason.
     Panic = 1,
     /// Go received a bad argument from Rust
     BadArgument = 2,
-    /// Ran out of gas while using the SDK (e.g. storage)
+    /// Ran out of gas while using the SDK (e.g. storage). This can come from the Cosmos SDK gas meter
+    /// (https://github.com/cosmos/cosmos-sdk/blob/v0.45.4/store/types/gas.go#L29-L32).
     OutOfGas = 3,
-    /// An error happened during normal operation of a Go callback, which should abort the contract
-    Other = 4,
+    /// Error while trying to serialize data in Go code (typically json.Marshal)
+    CannotSerialize = 4,
     /// An error happened during normal operation of a Go callback, which should be fed back to the contract
     User = 5,
+    /// An error type that should never be created by us. It only serves as a fallback for the i32 to GoError conversion.
+    Other = -1,
 }
 
-impl From<i32> for GoResult {
+impl From<i32> for GoError {
     fn from(n: i32) -> Self {
-        use GoResult::*;
         // This conversion treats any number that is not otherwise an expected value as `GoError::Other`
         match n {
-            0 => Ok,
-            1 => Panic,
-            2 => BadArgument,
-            3 => OutOfGas,
-            5 => User,
-            _ => Other,
+            0 => GoError::None,
+            1 => GoError::Panic,
+            2 => GoError::BadArgument,
+            3 => GoError::OutOfGas,
+            4 => GoError::CannotSerialize,
```

Comment:
You just added this one, right? Or was it missing before?

Reply:
The `Other` case with number 4 was previously used only once for the serialization failure. I think it's better to have that explicit. So what I did:
- Rename `Other` to `CannotSerialize` under number 4
- Create a new `Other` under number -1 that is not used